### ISSUE TITLE
Memory error at 14Gb with 25Gb memory

### ISSUE THREAD

**Issue by danny77277** [sokrypton/ColabFold#45]
Hello, 

I am having a memory error while trying to analyze my protein (sequence and error message attached below). I am using Colab Pro with 25Gb memory. The prediction works if I cut my protein in half, but I would like to analyze the full length protein if possible. Can I know if there is a way to get this to work?

Thanks!
Danny

**Protein sequence:**

EESAAPQVHLSILATTDIHANMMDYDYYSDKETADFGLARTAQLIQKHREQNPNTLLVDNGDLIQGNPLGEYAVKYQKDDIISGTKTHPIISVMNALKYDAGTLGNHEFNYGLDFLDGTIKGADFPIVNANVKTTSGENRYTPYVINEKTLIDENGNEQKVKVGYIGFVPPQIMTWDKKNLEGQVQVQDIVESANETIPKMKAEGADVIIALAHTGIEKQAQSSGAENAVFDLATKTKGIDAIISGHQHGLFPSAEYAGVAQFNVEKGTINGIPVVMPSSWGKYLGVIDLKLEKADGSWKVADSKGSIESIAGNVTSRNETVTNTIQQTHQNTLEYVRKPVGKTEADINSFFAQVKDDPSIQIVTDAQKWYAEKEMKDTEYKNLPILSAGAPFKAGGRNGANYYTNIPAGDLAIKNVGDLYLYDNTVQIVKLTGSEVKDWLEMSAGQFNQIDPAKGGDQALLNENFRSYNFDVIDGVTYQVDVTKPAKYNENGKVINADSSRIINLSYEGKPISPSQEFLVVTNNYRASGGGGFPHLTSDKIVHGSAVENRQVLMDYIIEQKTVNPKADNNWSIAPVSGTNLTFESSLLAKPFADKADDVAYVGKSANEGYGVYKLQFDDDSNPDPPKDGLWDLTVMHTNDTHAHLDDAARRMTKINEVRSETNHNILLDAGDVFSGDLYFTKWNGLADLKMMNMMGYDAMTFGNHEFDKGPTVLSDFLSGNSATVDPANRYHFEAPEFPIVSANVDVSNEPKLKSFVKKPQTFTAGEKKEAGIHPYILLDVDGEKVAVFGLTTEDTATTSSPGKSIVFNDAFETAQNTVKAIQEEEKVNKIIALTHIGHNRDLELAKKVKGIDLIIGGHTHTLVDKMEVVNNEEPTIVAQAKEYGQFLGRVDVAFDEKGVVQTDKSNLSVLPIDEHTEENPEAKQELDQFKNELEDVKNEKVGYTDVALDGQREHVRTKETNLGNFIADGMLAKAKEAAGARIAITNGGGIRAGIDKGDITLGEVLNVMPFGNTLYVADLTGKQIKEALEQGLSNVENGGGAFPQVAGIEYTFTLNNKPGHRVLEVKIESPNGDKVAINTDDTYRVATNNFVGAGGDGYSVFTEASHGEDLGYVDYEIFTEQLKKLGNKVSPKVEGRIKEVFLPTKQKDGSWTLDEDKFAIYAKNANTPFVYYGIHEGSQEKPINLKVKKDQVKLLKERESDPSLTMFNYWYSMKMPMANLKTADTAIGIKSTGELDVSLSDVYDFTVKQKGKEIKSFKEPVQLSLRMFDIEEAHNPAIYHVDRKKKAFTKTGHGSVDDDMVTGYTNHFSEYTILNSGSNNKPPAFPSDQPTGGDDGNHGGGSDKPGGKQPTDGNGGNDTPPGTQPTNGSGGNGSGGSGTDGPAGGLLPDT

**Error message:**

running model_1
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-5-af48741e914e> in <module>()
     50                          model_params=model_params, use_model=use_model,
---> 51                          do_relax=use_amber)

13 frames
UnfilteredStackTrace: RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/interpreters/xla.py in _execute_compiled(compiled, avals, handlers, kept_var_idx, *args)
    910           for i, x in enumerate(args)
    911           if x is not token and i in kept_var_idx))
--> 912   out_bufs = compiled.execute(input_bufs)
    913   check_special(xla_call_p.name, out_bufs)
    914   return [handler(*bs) for handler, bs in zip(handlers, _partition_outputs(avals, out_bufs))]

RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

**Comment by sokrypton** [comment]
How did you determine the amount of GPU memory available?
Try adding a new cell and type `!nvidia-smi`

**Comment by danny77277** [comment]
Thanks! I thought the memory was the number at the top right (where the connect button is). It said GPU and 25Gb. I also tried adding the command you suggested. Looks like the GPU memory is actually 16Gb. I got the same error later. Looks like the sequence is a bit too long for the analysis.

Danny


Sun Sep 19 14:50:51 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.63.01    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

The process stopped with the same error later:

running model_1
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-6-af48741e914e> in <module>()
     50                          model_params=model_params, use_model=use_model,
---> 51                          do_relax=use_amber)

13 frames
UnfilteredStackTrace: RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/interpreters/xla.py in _execute_compiled(compiled, avals, handlers, kept_var_idx, *args)
    910           for i, x in enumerate(args)
    911           if x is not token and i in kept_var_idx))
--> 912   out_bufs = compiled.execute(input_bufs)
    913   check_special(xla_call_p.name, out_bufs)
    914   return [handler(*bs) for handler, bs in zip(handlers, _partition_outputs(avals, out_bufs))]

RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

**Comment by truatpasteurdotfr** [comment]
try setting `TF_FORCE_UNIFIED_MEMORY=1,XLA_PYTHON_CLIENT_MEM_FRACTION=4`

**Comment by danny77277** [comment]
Thanks for the suggestion! I tried to add the setting (included below) and got the same error during prediction (16Gb GPU).

#@title Input protein sequence, then hit `Runtime` -> `Run all`
from google.colab import files
import os
import os.path
import re
import hashlib

**os.environ['TF_FORCE_UNIFIED_MEMORY']='1'
os.environ['XLA_PYTHON_CLIENT_MEM_FRACTION']='.5'**
def add_hash(x,y):
  return x+"_"+hashlib.sha1(y.encode()).hexdigest()[:5]

query_sequence = 'EESAAPQVHLSILATTDIHANMMDYDYYSDKETADFGLARTAQLIQKHREQNPNTLLVDNGDLIQGNPLGEYAVKYQKDDIISGTKTHPIISVMNALKYDAGTLGNHEFNYGLDFLDGTIKGADFPIVNANVKTTSGENRYTPYVINEKTLIDENGNEQKVKVGYIGFVPPQIMTWDKKNLEGQVQVQDIVESANETIPKMKAEGADVIIALAHTGIEKQAQSSGAENAVFDLATKTKGIDAIISGHQHGLFPSAEYAGVAQFNVEKGTINGIPVVMPSSWGKYLGVIDLKLEKADGSWKVADSKGSIESIAGNVTSRNETVTNTIQQTHQNTLEYVRKPVGKTEADINSFFAQVKDDPSIQIVTDAQKWYAEKEMKDTEYKNLPILSAGAPFKAGGRNGANYYTNIPAGDLAIKNVGDLYLYDNTVQIVKLTGSEVKDWLEMSAGQFNQIDPAKGGDQALLNENFRSYNFDVIDGVTYQVDVTKPAKYNENGKVINADSSRIINLSYEGKPISPSQEFLVVTNNYRASGGGGFPHLTSDKIVHGSAVENRQVLMDYIIEQKTVNPKADNNWSIAPVSGTNLTFESSLLAKPFADKADDVAYVGKSANEGYGVYKLQFDDDSNPDPPKDGLWDLTVMHTNDTHAHLDDAARRMTKINEVRSETNHNILLDAGDVFSGDLYFTKWNGLADLKMMNMMGYDAMTFGNHEFDKGPTVLSDFLSGNSATVDPANRYHFEAPEFPIVSANVDVSNEPKLKSFVKKPQTFTAGEKKEAGIHPYILLDVDGEKVAVFGLTTEDTATTSSPGKSIVFNDAFETAQNTVKAIQEEEKVNKIIALTHIGHNRDLELAKKVKGIDLIIGGHTHTLVDKMEVVNNEEPTIVAQAKEYGQFLGRVDVAFDEKGVVQTDKSNLSVLPIDEHTEENPEAKQELDQFKNELEDVKNEKVGYTDVALDGQREHVRTKETNLGNFIADGMLAKAKEAAGARIAITNGGGIRAGIDKGDITLGEVLNVMPFGNTLYVADLTGKQIKEALEQGLSNVENGGGAFPQVAGIEYTFTLNNKPGHRVLEVKIESPNGDKVAINTDDTYRVATNNFVGAGGDGYSVFTEASHGEDLGYVDYEIFTEQLKKLGNKVSPKVEGRIKEVFLPTKQKDGSWTLDEDKFAIYAKNANTPFVYYGIHEGSQEKPINLKVKKDQVKLLKERESDPSLTMFNYWYSMKMPMANLKTADTAIGIKSTGELDVSLSDVYDFTVKQKGKEIKSFKEPVQLSLRMFDIEEAHNPAIYHVDRKKKAFTKTGHGSVDDDMVTGYTNHFSEYTILNSGSNNKPPAFPSDQPTGGDDGNHGGGSDKPGGKQPTDGNGGNDTPPGTQPTNGSGGNGSGGSGTDGPAGGLLPDT' #@param {type:"string"}



running model_1
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-10-af48741e914e> in <module>()
     50                          model_params=model_params, use_model=use_model,
---> 51                          do_relax=use_amber)

13 frames
UnfilteredStackTrace: RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/interpreters/xla.py in _execute_compiled(compiled, avals, handlers, kept_var_idx, *args)
    910           for i, x in enumerate(args)
    911           if x is not token and i in kept_var_idx))
--> 912   out_bufs = compiled.execute(input_bufs)
    913   check_special(xla_call_p.name, out_bufs)
    914   return [handler(*bs) for handler, bs in zip(handlers, _partition_outputs(avals, out_bufs))]

RuntimeError: Resource exhausted: Out of memory while trying to allocate 14268435552 bytes.

**Comment by zach-hensel** [comment]
GPU memory is 16 GB... the high RAM colab pro instances have more system RAM and and better odds of getting better GPUs. Setting 3 environment variables (per the alphafold docker image) works for me locally; I also just tested and this works on colab.

The 4.0 and not 0.5 (or 1.0 or sometimes 2.0) settings was important for me locally.

Here's what's in my .bashrc to work locally using https://github.com/YoshitakaMo/localcolabfold -
```
export NVIDIA_VISIBLE_DEVICES='all'
export TF_FORCE_UNIFIED_MEMORY='1'
export XLA_PYTHON_CLIENT_MEM_FRACTION='4.0'
```
And this worked on colabfold -- I added this cell to the top of the advanced notebook.
```
!nvidia-smi
%env TF_FORCE_UNIFIED_MEMORY=all
%shell echo $NVIDIA_VISIBLE_DEVICES
%env TF_FORCE_UNIFIED_MEMORY=1
%shell echo $TF_FORCE_UNIFIED_MEMORY
%env XLA_PYTHON_CLIENT_MEM_FRACTION=4.0
%shell echo $XLA_PYTHON_CLIENT_MEM_FRACTION
```
The 1460-residue (1 recycle only) test prediction on colab peaked at 15.9 GB GPU and 9.5 GB system RAM... not a ton of headroom to go for larger complexes, and it is **slow**. The free credit on google cloud compute is enough to see what more GPU RAM can do for you.
